### PR TITLE
Programmatic markup generation

### LINKED ISSUE #155: Render tagging programatically without templates

### PR DESCRIPTION
**Pull request by chenliangde**
Closes #155

### CODE REVIEW

File: classes/blocks/NostoCustomerTagging.php [hunk 1]
```diff
@@ -41,12 +41,9 @@ public static function get(NostoTagging $module)
         $hcid = $cid ? hash(NostoTagging::VISITOR_HASH_ALGO, $cid) : '';
 
         $nostoCustomer = NostoCustomer::loadData(Context::getContext()->customer);
-        Context::getContext()->smarty->assign(array(
```
Comment (mridang) [inline classes/blocks/NostoCustomerTagging.php]: Check if null just like `NostoCart`

File: classes/blocks/NostoVariationTagging.php [hunk 2]
```diff
@@ -39,10 +39,7 @@ public static function get(NostoTagging $module)
         }
 
         $priceVariation = NostoVariation::loadData();
-        Context::getContext()->smarty->assign(array(
```
Comment (mridang) [inline classes/blocks/NostoVariationTagging.php]: Check if null just like `NostoCart`

File: classes/blocks/NostoBrandTagging.php [hunk 3]
```diff
@@ -40,10 +40,7 @@ public static function get(NostoTagging $module)
         }
 
         $nostoBrand = NostoBrand::loadData($brand);
-        Context::getContext()->smarty->assign(array(
```
Comment (mridang) [inline classes/blocks/NostoBrandTagging.php]: Check if null just like `NostoCart`

File: classes/blocks/NostoCategoryTagging.php [hunk 4]
```diff
@@ -47,12 +47,8 @@ public static function get(NostoTagging $module)
             $idCategory = Context::getContext()->cookie->last_visited_category;
             $category = NostoCategory::loadId($idCategory, Context::getContext()->language->id);
         }
-
         $nostoCategory = NostoCategory::loadData(Context::getContext(), $category);
-        Context::getContext()->smarty->assign(array(
```
Comment (mridang) [inline classes/blocks/NostoCategoryTagging.php]: Check if null just like `NostoCart`

File: classes/blocks/NostoOrderTagging.php [hunk 5]
```diff
@@ -40,10 +40,6 @@ public static function get(NostoTagging $module)
 
         $nostoOrder = NostoOrder::loadData(Context::getContext(), $order);
 
-        Context::getContext()->smarty->assign(array(
```
Comment (mridang) [inline classes/blocks/NostoOrderTagging.php]: Check if null just like `NostoCart`

File: classes/blocks/NostoCategoryTagging.php [hunk 6]
```diff
@@ -47,12 +47,8 @@ public static function get(NostoTagging $module)
...
-            'nosto_category' => $nostoCategory,
-        ));
 
-        return $module->render('views/templates/hook/category-footer_category-tagging.tpl');
+        return $nostoCategory ? $nostoCategory->toHtml() : '';
```
Comment (mridang) [inline classes/blocks/NostoCategoryTagging.php]: Should this return null too?

File: classes/blocks/NostoSearchTagging.php [hunk 7]
```diff
@@ -40,10 +40,7 @@ public static function get(NostoTagging $module)
         }
 
         $nostoQuery = NostoSearch::loadData($searchTerm);
-        Context::getContext()->smarty->assign(array(
-            'nosto_search' => $nostoQuery,
-        ));
 
-        return $module->render('views/templates/hook/top_search-tagging.tpl');
+        return $nostoQuery->toHtml();
```
Comment (mridang) [inline classes/blocks/NostoSearchTagging.php]: Use ternary to check if null

File: classes/blocks/NostoVariationTagging.php [hunk 8]
```diff
@@ -39,10 +39,7 @@ public static function get(NostoTagging $module)
...
-            'nosto_price_variation' => $priceVariation
-        ));
 
-        return $module->render('views/templates/hook/top_price_variation-tagging.tpl');
+        return $priceVariation->toHtml();
```
Comment (mridang) [inline classes/blocks/NostoVariationTagging.php]: Use ternary to check if null

File: classes/blocks/NostoBrandTagging.php [hunk 9]
```diff
@@ -40,10 +40,7 @@ public static function get(NostoTagging $module)
...
-            'nosto_brand' => $nostoBrand,
-        ));
 
-        return $module->render('views/templates/hook/manufacturer-footer_brand-tagging.tpl');
+        return $nostoBrand->toHtml();
```
Comment (mridang) [inline classes/blocks/NostoBrandTagging.php]: Use ternary to check if null

**Comment by mridang** [comment]
Also, you have a typo in your branch name. `feature/remove-temlate`

File: classes/models/NostoProduct.php [hunk 10]
```diff
@@ -242,23 +243,26 @@ public static function getListPriceInclTax(Product $product, Context $context, C
      *
      * @param Product $product the product model.
      * @param int $id_lang for which language ID to fetch the product tags.
-     * @return array the built tags.
+     * @return StringCollection the built tags.
      */
     protected static function buildTags(Product $product, $id_lang)
     {
-        $tags = array();
+        $tagsCollection = new StringCollection('tag1', 'tag');
```
Comment (mridang) [inline classes/models/NostoProduct.php]: Move to SDK

**Comment by chenliangde** [comment]
An other PR has been created for fixing the branch name:
https://github.com/Nosto/nosto-prestashop/pull/168

**Pull request closed without merging**